c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,338 @@
+  gs.add<&Ftrl::get_model, &Ftrl::set_model>("model",
+    "Frame having two columns, i.e. `z` and `n`, and `d` rows,\n"
+    "where `d` is a number of bins set for modeling. Both column types\n"
+    "must be `FLOAT64`.\n"
+    "NB: as the model trains, this frame will be changed in-place.\n");
+
+  gs.add<&Ftrl::get_a, &Ftrl::set_a>("a", "`alpha` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_b, &Ftrl::set_b>("b", "`beta` in per-coordinate learning rate formula.\n");
+  gs.add<&Ftrl::get_l1, &Ftrl::set_l1>("l1", "L1 regularization parameter.\n");
+  gs.add<&Ftrl::get_l2, &Ftrl::set_l2>("l2", "L2 regularization parameter.\n");
+  gs.add<&Ftrl::get_d, &Ftrl::set_d>("d", "Number of bins to be used after the hashing trick.\n");
+  gs.add<&Ftrl::get_n_epochs, &Ftrl::set_n_epochs>("n_epochs", "Number of epochs to train for.\n");
+  gs.add<&Ftrl::get_inter, &Ftrl::set_inter>("inter", "If feature interactions to be used or not.\n");
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type",
+    "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"df_train"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    A new Frame of shape `(d, 2)` containing `z` and `n` model coefficients.
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+  DataTable* dt_train = args[0].to_frame();
+  fm->fit(dt_train);
+}
+
+
+PKArgs Ftrl::Type::args_predict(1, 0, 0, false, false, {"df_test"}, "predict",
+R"(predict(self, frame)
+--
+
+Make predictions for a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame of shape `(nrows, ncols)` to make predictions for. It must have one 
+    column less than the training dataset.
+
+Returns
+----------
+    A new `Frame` of shape `(nrows, 1)` with a prediction for each row.
+)");
+
+
+oobj Ftrl::predict(const PKArgs& args) {
+  if (fm->get_model() == nullptr) {
+    throw ValueError() << "There is no trained model in place.\n"
+                       << "To make predictions, the model must be either trained first, or "
+                       << "set.";
+  }
+
+  if (!args[0].is_frame()) {
+    throw TypeError() << "argument must be a frame, not "
+        << args[0].typeobj();
+  }
+
+  DataTable* dt_test = args[0].to_frame();
+  DataTable* dt_target = fm->predict(dt_test).release();
+  py::Frame* df_target = py::Frame::from_datatable(dt_target);
+
+  return df_target;
```

Comment:
I suspect this creates a memory leak when `df_target` is converted to `py::oobj`. Use `py::oobj::from_new_reference(df_target)` instead.

Reply:
Oops. I'm also using the same thing in `Aggregator`. Why this will create a memory leak?

Reply:
Whenever you're dealing with `PyObject*`s, you have to be really careful that its ref-count is incremented/decremented in equal amounts. This is why we have "python/_all.h" mini-library to begin with. And of course, it is my fault that `py::Frame` is visible externally as `PyObject*` instead of `py::oobj`. 

Anyways, the reason for the leak is the following: `py::Frame::from_datatable` creates a new `py::Frame*` object, which derives from `PyObject*`. Naturally, this object is created with ref-count of 1. When you return that object, the compiler silently converts `df_target` into `py::oobj((PyObject*)df_target)` to match the return type. And the constructor of `py::oobj()` increments the ref-count too (unless you construct via `py::oobj::from_new_reference()`). The end result is that the returned frame will have ref-count of 2. 

Later when the frame is deleted, its ref-count will drop to 1, leaving it as an orphaned pointer. I don't know whether the garbage collector will be able to clean this up, so I wouldn't trust that.

Reply:
Thanks for the explanation. From what I see `py::oobj::from_new_reference` takes `PyObject*` as an argument, and we have `DataTable*`. I see that in other parts of the code you're doing `py::oobj::from_new_reference(py::Frame::from_datatable(dt))`, is it what you suggest and won't this create the same kind of memory leak?